driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
       }
     }
 
-    if (connections[shardId].isEmpty()) {
-      if (host.convictionPolicy.canReconnectNow()) {
-        if (connectionsPerShard == 0) {
-          maybeSpawnNewConnection(shardId);
-        } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
-          for (int i = 0; i < connectionsPerShard; i++) {
-            // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
-            // protect against creating connection in excess of core too quickly
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
+    int shId = shardId;
+    do {
+    	synchronized(connections[shId]) {
```

Comment:
This is wrong. This method is supposed to be non-blocking (it returns future) and this is a synchronous lock.

Reply:
The method is non-blocking. This synchronized block protects concurrent access to a non thread-safe structure (ArrayList). It doesn't block the entire method, just this snippet. Concurrent access to non thread-safe structures lead to unpredictable behavior (https://www.amazon.com/Taming-Java-Threads-Allen-Holub/dp/1893115100)

Reply:
Are you serious? 
> It doesn't block the entire method, just this snippet.

What does that even mean?

As an experienced Java architect you surely know that ["All synchronized blocks synchronized on the same object can only have one thread executing inside them at the same time. All other threads attempting to enter the synchronized block **are blocked** until the thread inside the synchronized block exits the block"](http://tutorials.jenkov.com/java-concurrency/synchronized.html)

Reply:
We clearly don't talk about the same thing. The issue I'm raising is that synchronized block is like a mutex. If one thread acquires it, all other threads have to wait (hence block) until the mutex is released. This means a method with a synchronized block is blocking.

Reply:
Yes, that's the definition of a thread-safe block. So, what are you advocating? To avoid using any kind of locks and synchronization mechanisms? To begin with, "connections" should not be an ArrayList.

Reply:
I'm saying that the code before was thread-safe and lock-free. We should keep it that way. I'm not saying we have to avoid all synchronization mechanisms - only those that block.

Why do you think connections shouldn't be ArrayList?